hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/handler/AssignRegionHandler.java

```diff
@@ -105,50 +102,54 @@ public void process() throws IOException {
         // closing process.
         long backoff = retryCounter.getBackoffTimeAndIncrementAttempts();
-        LOG.info(
-          "Receiving OPEN for the region:{}, which we are trying to close, try again after {}ms",
+        LOG.info("Receiving OPEN for {} which we are trying to CLOSE, try again after {}ms",
           regionName, backoff);
         rs.getExecutorService().delayedSubmit(this, backoff, TimeUnit.MILLISECONDS);
       }
       return;
     }
     LOG.info("Open {}", regionName);
     HRegion region;
+    // Start a try. Above we registered region with rs.getRegionsInTransitionInRS. Need to clear
+    // it on success or error on the end.
     try {
-      TableDescriptor htd =
-        tableDesc != null ? tableDesc : rs.getTableDescriptors().get(regionInfo.getTable());
-      if (htd == null) {
-        throw new IOException("Missing table descriptor for " + regionName);
+      try {
+        TableDescriptor htd =
+          tableDesc != null ? tableDesc : rs.getTableDescriptors().get(regionInfo.getTable());
+        if (htd == null) {
+          throw new IOException("Missing table descriptor for " + regionName);
+        }
+        // pass null for the last parameter, which used to be a CancelableProgressable, as now the
+        // opening can not be interrupted by a close request any more.
+        region = HRegion
+          .openHRegion(regionInfo, htd, rs.getWAL(regionInfo), rs.getConfiguration(), rs, null);
+      } catch (IOException e) {
+        cleanUpAndReportFailure(e);
+        return;
       }
-      // pass null for the last parameter, which used to be a CancelableProgressable, as now the
-      // opening can not be interrupted by a close request any more.
-      region = HRegion.openHRegion(regionInfo, htd, rs.getWAL(regionInfo), rs.getConfiguration(),
-        rs, null);
-    } catch (IOException e) {
-      cleanUpAndReportFailure(e);
-      return;
-    }
-    rs.postOpenDeployTasks(new PostOpenDeployContext(region, openProcId, masterSystemTime));
-    rs.addRegion(region);
-    LOG.info("Opened {}", regionName);
-    // Cache the open region procedure id after report region transition succeed.
-    rs.finishRegionProcedure(openProcId);
-    Boolean current = rs.getRegionsInTransitionInRS().remove(regionInfo.getEncodedNameAsBytes());
-    if (current == null) {
-      // Should NEVER happen, but let's be paranoid.
-      LOG.error("Bad state: we've just opened a region that was NOT in transition. Region={}",
-        regionName);
-    } else if (!current) {
-      // Should NEVER happen, but let's be paranoid.
-      LOG.error("Bad state: we've just opened a region that was closing. Region={}", regionName);
+      rs.postOpenDeployTasks(new PostOpenDeployContext(region, openProcId, masterSystemTime));
+      rs.addRegion(region);
+      LOG.info("Opened {}", regionName);
+    } finally {
+      // Remove from regionsInTransitionRS whether success or failure.
```

Comment:
this is the crux of your test fix -- that we clean up after ourselves properly?

Reply:
y